### PR TITLE
dbeaver/pro#1905 Incorrect characters in databricks tables

### PR DESCRIPTION
**Pull request by serjiokov**
Fixed specification of file.encoding property require for correct string processing.
Provided mechanism in order to change default encoding value

### CODE REVIEW

**Comment by serjiokov** [comment]
Since now, user able to specify require encoding type
image

**Review by E1izabeth — COMMENTED** [review]
Why the setting is called `File encoding`? Do I correctly understand, that it works not only with files since you're fixing issue with result set?

**Comment by serjiokov** [comment]
I do name as File Encoding because it has correlation to original property.
The **file.encoding** value influences the default charset used by the JVM among other things

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/utils/GeneralUtils.java [hunk 1]
```diff
@@ -829,4 +847,56 @@ public static UUID getMixedEndianUUIDFromBytes(byte[] bytes) {
         target.rewind();
         return new UUID(target.getLong(), target.getLong());
     }
+
+    /**
+     * Find index of item in set
+     */
+    public static <T> int getIndexInSet(Set<T> set, @Nullable T value) {
```
Comment (serge-rider) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/utils/GeneralUtils.java]: Do not add common util methods in GeneralUtils.

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/utils/GeneralUtils.java [hunk 2]
```diff
@@ -829,4 +847,56 @@ public static UUID getMixedEndianUUIDFromBytes(byte[] bytes) {
...
+        List<String> charsetList = new ArrayList<>();
+        charsetList.add(String.format("%s (%s)", "Default encoding", systemFileEncoding));
+        charsetList.addAll(Charset.availableCharsets().keySet());
+        return charsetList;
+    }
+
+    /**
+     * Return index of charset specification
+     */
+    public static int getIndexOfAvailableCharset(String encoding) {
```
Comment (serge-rider) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/utils/GeneralUtils.java]: This function doesn't make sense

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/utils/GeneralUtils.java [hunk 3]
```diff
@@ -829,4 +847,56 @@ public static UUID getMixedEndianUUIDFromBytes(byte[] bytes) {
...
+        return availableCharsets().indexOf(encoding);
+    }
+
+    /**
+     * Return charset specification by index
+     */
+    public static String getCharsetByIndex(int index) {
```
Comment (serge-rider) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/utils/GeneralUtils.java]: This too. Just use encoding name from combobox text

**Review by serge-rider — CHANGES_REQUESTED** [review]

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/utils/GeneralUtils.java [hunk 4]
```diff
@@ -114,6 +115,23 @@ public static String getDefaultConsoleEncoding() {
...
+        } else {
+            setDeafaultFileEncoding();
+        }
+    }
+
+    public static void setDeafaultFileEncoding() {
+        System.setProperty(StandardConstants.ENV_FILE_ENCODING, UTF8_ENCODING);
+    }
+
+    public static String getFileEncoding() {
```
Comment (serge-rider) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/utils/GeneralUtils.java:130]: This function doesn't make sense

**Review by serge-rider — CHANGES_REQUESTED** [review]

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/utils/GeneralUtils.java (hunk 4 above)
Comment (serjiokov) [reply plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/utils/GeneralUtils.java:130]: Done

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/utils/GeneralUtils.java (hunk 3 above)
Comment (serjiokov) [reply plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/utils/GeneralUtils.java]: Done

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/utils/GeneralUtils.java (hunk 2 above)
Comment (serjiokov) [reply plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/utils/GeneralUtils.java]: Done

**Comment by serjiokov** [comment]
image

**Review by serge-rider — APPROVED** [review]

**Comment by E1izabeth** [comment]
The issues that might be solved with this PR (_**we need to check**_):
https://github.com/dbeaver/dbeaver/issues/20784
https://github.com/dbeaver/dbeaver/issues/20392
https://github.com/dbeaver/dbeaver/issues/20054
https://github.com/dbeaver/dbeaver/issues/19631
https://github.com/dbeaver/dbeaver/issues/9890
https://github.com/dbeaver/dbeaver/issues/8875
https://github.com/dbeaver/dbeaver/issues/6858

**Review by E1izabeth — COMMENTED** [review]
Won't this preference change will take effect only after DBeaver restart?
If yes we need to add such info label then (like in Regional settings group). And then it would be better to move this option to the upper group to prevent duplication of the info label.

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/utils/GeneralUtils.java [hunk 5]
```diff
@@ -114,6 +114,23 @@ public static String getDefaultConsoleEncoding() {
...
+    public static void setFileEncoding(String string) {
+        Charset charset = Charset.availableCharsets().get(string);
+        if (charset != null) {
+            System.setProperty(StandardConstants.ENV_FILE_ENCODING, charset.name());
+        } else {
+            setDeafaultFileEncoding();
+        }
+    }
+
+    public static void setDeafaultFileEncoding() {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/utils/GeneralUtils.java:126]: Typo

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/app/DBPApplication.java [hunk 6]
```diff
@@ -102,5 +102,14 @@ default void beforeWorkspaceInitialization() {}
      */
     long getApplicationStartTime();
 
+    /**
+     * Store system encoding specification
+     */
+    public void storeSystemEncoding(String encoding);
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/app/DBPApplication.java:108]: Please annotate public interface methods with nullability annotations

File: plugins/org.jkiss.dbeaver.core/src/org/jkiss/dbeaver/ui/preferences/PrefPageDatabaseUserInterface.java [hunk 7]
```diff
@@ -309,4 +341,14 @@ public IAdaptable getElement()
...
     {
     }
+
+    /**
+     * Return available charsets with first item of system encoding
+     */
+    public static List<String> availableCharsets() {
+        List<String> charsetList = new ArrayList<>();
+        charsetList.add(String.format("%s (%s)", "Default encoding", BaseApplicationImpl.getInstance().getSystemEncoding()));
+        charsetList.addAll(Charset.availableCharsets().keySet());
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.core/src/org/jkiss/dbeaver/ui/preferences/PrefPageDatabaseUserInterface.java:351]: I expect this set to be ordered chaotically. Maybe it's better to sort it alphabetically?

File: plugins/org.jkiss.dbeaver.core/src/org/jkiss/dbeaver/ui/preferences/PrefPageDatabaseUserInterface.java [hunk 8]
```diff
@@ -74,11 +77,13 @@ public class PrefPageDatabaseUserInterface extends AbstractPrefPage implements I
 
     @Nullable
     private Combo clientTimezone;
+    private Combo cmbDefaultEncoding;
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.core/src/org/jkiss/dbeaver/ui/preferences/PrefPageDatabaseUserInterface.java:80]:
```suggestion
    private Combo defaultEncodingCombo;
```

We don't really use abbreviations such as cmb, btn, etc.

**Pull request closed without merging**